File: backend/app/utils/enums.py
"""backend/app/utils/enums.py — Enumerations for incident types, severity levels, statuses, and user roles used throughout E·OVR.

GAHAR migration changes:
  - Severity: added Catastrophic as the new highest tier (replaces the JCI 3-level Major/Moderate/Minor scale).
  - Probability: replaced Low/Medium/High (JCI 3-level) with
    Frequent/Occasional/Uncommon/Remote (GAHAR 4-level SAC scale).
  - ErrorClassification: added WorkplaceViolence and Environmental (moved here
    from the broken Enum() factory block in incident.py — fixes the 500 on list).

EN fields migration:
  - FacilityType values remain Arabic strings (stored in DB as Arabic) so that
    existing incidents and facility documents are not invalidated.
  - FACILITY_TYPE_EN provides a mapping from Arabic value → English label for
    display purposes (API responses, frontend dropdowns).
"""

from enum import Enum


class IncidentStatus(str, Enum):
    """Lifecycle state of an incident report, from initial submission through resolution."""

    Created = "Created"
    InProgress = "InProgress"
    Evaluating = "Evaluating"
    MoreInfoNeeded = "MoreInfoNeeded"
    ActionTaken = "ActionTaken"
    Completed = "Completed"


class Severity(str, Enum):
    """Clinical or operational severity of the incident.

    GAHAR SAC scale (4 levels, highest → lowest):
      Catastrophic — patient death or permanent harm / sentinel event
      Major        — serious but non-permanent harm
      Moderate     — significant but recoverable harm
      Minor        — minimal or no patient harm
    """

    Catastrophic = "Catastrophic"
    Major = "Major"
    Moderate = "Moderate"
    Minor = "Minor"


class Probability(str, Enum):
    """Likelihood of a similar incident recurring.

    GAHAR SAC scale (4 levels, most → least likely):
      Frequent   — expected to occur several times per year
      Occasional — expected to occur once per year or less
      Uncommon   — may occur once every 2–5 years
      Remote     — unlikely; may occur once in > 5 years
    """

    Frequent = "Frequent"
    Occasional = "Occasional"
    Uncommon = "Uncommon"
    Remote = "Remote"


class ErrorClassification(str, Enum):
    """Consolidated 14-category incident taxonomy.

    Replaces both the old 8-class GAHAR-era taxonomy (Medication, Administrative,
    Clinical, Equipment, InfectionControl, Falls, Documentation, Behavior) and the
    separate IncidentMainCategory field.  Legacy values are coerced in
    IncidentInDB._coerce_legacy_error_classification.
    """

    PatientSafety               = "PatientSafety"
    MedicationSafety            = "MedicationSafety"
    InfectionPrevention         = "InfectionPrevention"
    BloodTransfusion            = "BloodTransfusion"
    Laboratory                  = "Laboratory"
    RadiologyDiagnostic         = "RadiologyDiagnostic"
    MedicalEquipment            = "MedicalEquipment"
    FacilityEnvironmental       = "FacilityEnvironmental"
    OccupationalHealthStaff     = "OccupationalHealthStaff"
    Security                    = "Security"
    InformationTechnology       = "InformationTechnology"
    AdministrativeProcess       = "AdministrativeProcess"
    PatientExperienceComplaints = "PatientExperienceComplaints"
    FireDisaster                = "FireDisaster"


class EventType(str, Enum):
    """Classification of the reported event according to its outcome and potential impact."""

    IncidentEvent = "IncidentEvent"
    NearMiss = "NearMiss"
    AdverseEvent = "AdverseEvent"
    SignificantEvent = "SignificantEvent"
    SentinelEvent = "SentinelEvent"


class ActionStatus(str, Enum):
    """Current completion state of a corrective or preventive action item."""

    Pending = "Pending"
    InProgress = "InProgress"
    Completed = "Completed"


class ReporterType(str, Enum):
    """Indicates whether the person submitting the report is a patient or a staff member."""

    patient = "patient"
    staff = "staff"


class UserRole(str, Enum):
    """Access-control role assigned to a user account, governing what the user can view and do."""

    patient = "patient"
    staff = "staff"
    quality_admin = "quality_admin"
    administration_manager = "administration_manager"
    governorate_manager = "governorate_manager"
    top_management = "top_management"


class FacilityType(str, Enum):
    """Arabic label for the type of healthcare facility.

    Values are stored in Arabic in MongoDB so that existing incidents and
    facility documents remain valid.  Use FACILITY_TYPE_EN for English display.
    """

    hospital = "مستشفى"
    center = "مركز"
    unit = "وحدة"


# English display labels for FacilityType — keyed by the Arabic stored value.
# Use this wherever you need to show the English name without touching the DB schema.
# Example: FACILITY_TYPE_EN.get(facility.facility_type.value, facility.facility_type.value)
FACILITY_TYPE_EN: dict[str, str] = {
    "مستشفى": "Hospital",
    "مركز":   "Health Center",
    "وحدة":   "Health Unit",
}


class EventDiscoveryMethod(str, Enum):
    """How the event was first identified / brought to attention."""

    SelfReported           = "SelfReported"
    SupervisorReported     = "SupervisorReported"
    PatientComplaint       = "PatientComplaint"
    AuditFinding           = "AuditFinding"
    IncidentInvestigation  = "IncidentInvestigation"
    Other                  = "Other"


class NCCMERPCategory(str, Enum):
    """NCC MERP Index classification for medication error severity (A–I)."""

    A = "A"  # Circumstances with capacity to cause error — no error occurred
    B = "B"  # Error occurred but did not reach the patient
    C = "C"  # Error reached patient, no harm
    D = "D"  # Error reached patient, required monitoring/intervention to confirm no harm
    E = "E"  # Temporary harm requiring intervention
    F = "F"  # Temporary harm requiring initial or prolonged hospitalisation
    G = "G"  # Permanent patient harm
    H = "H"  # Intervention required to sustain life
    I = "I"  # Patient death


class MedicationErrorStage(str, Enum):
    """Stage in the medication use process at which the error occurred."""

    Prescribing    = "Prescribing"
    Transcribing   = "Transcribing"
    Dispensing     = "Dispensing"
    Preparation    = "Preparation"
    Administration = "Administration"
    Monitoring     = "Monitoring"


class AIProvider(str, Enum):
    """External AI/LLM provider used for automated analysis features; 'none' disables AI."""

    none = "none"
    openai = "openai"
    anthropic = "anthropic"
    google = "google"
    custom = "custom"


class ModelType(str, Enum):
    """Functional category of a machine-learning model registered in the AI model registry."""

    classification = "classification"
    similarity = "similarity"
    risk_scoring = "risk_scoring"
    signal_detection = "signal_detection"

File: frontend/src/utils/enums.js
/**
 * frontend/src/utils/enums.js
 *
 * Frontend enum constants — mirrors backend app/utils/enums.py.
 * Arrays drive <select> dropdowns; keyed objects drive badge/role display.
 *
 * Occurrence-location update:
 *   Added HOSPITAL_LOCATION_OPTIONS and OTHER_FACILITY_LOCATION_OPTIONS
 *   to power the facility-type-aware location dropdown in NewIncidentForm.
 *   Slugs are persisted to the DB; labels come from incidents.occurrence_location.*
 *   in the i18n translation files.
 */

// ── Incident Status — keyed object ────────────────────────────────────────────
// StatusBadge uses INCIDENT_STATUSES[status] → { label, color, bg }.
// Colors match STATUS_PILL in IncidentDetail.jsx for visual consistency.
export const INCIDENT_STATUSES = {
  Created:        { label: 'Created',          color: '#374151', bg: '#F3F4F6' },
  InProgress:     { label: 'In Progress',      color: '#1E40AF', bg: '#DBEAFE' },
  Evaluating:     { label: 'Evaluating',       color: '#92400E', bg: '#FEF3C7' },
  MoreInfoNeeded: { label: 'More Info Needed', color: '#B45309', bg: '#FEF3C7' },
  ActionTaken:    { label: 'Action Taken',     color: '#5B21B6', bg: '#EDE9FE' },
  Completed:      { label: 'Completed',        color: '#065F46', bg: '#D1FAE5' },
};

// ── User Roles — keyed object ─────────────────────────────────────────────────
// Sidebar uses USER_ROLES[role] → { label, tier }.
// Mirrors UserRole in enums.py; tiers reflect access-control hierarchy.
export const USER_ROLES = {
  patient:                { label: 'Patient',                tier: 1 },
  staff:                  { label: 'Staff',                  tier: 2 },
  quality_admin:          { label: 'Quality Admin',          tier: 3 },
  administration_manager: { label: 'Administration Manager', tier: 4 },
  governorate_manager:    { label: 'Governorate Manager',    tier: 4 },
  top_management:         { label: 'Top Management',         tier: 5 },
};

// ── Error Classification ──────────────────────────────────────────────────────
// Mirrors ErrorClassification in enums.py (14-category consolidated taxonomy)
export const ERROR_CLASSIFICATIONS = [
  'PatientSafety',
  'MedicationSafety',
  'InfectionPrevention',
  'BloodTransfusion',
  'Laboratory',
  'RadiologyDiagnostic',
  'MedicalEquipment',
  'FacilityEnvironmental',
  'OccupationalHealthStaff',
  'Security',
  'InformationTechnology',
  'AdministrativeProcess',
  'PatientExperienceComplaints',
  'FireDisaster',
];

// ── Event Type ────────────────────────────────────────────────────────────────
// Mirrors EventType in enums.py
export const EVENT_TYPES = [
  'IncidentEvent',
  'NearMiss',
  'AdverseEvent',
  'SignificantEvent',
  'SentinelEvent',
];

// ── Severity ──────────────────────────────────────────────────────────────────
// Mirrors Severity in enums.py (GAHAR SAC 4-level scale, highest → lowest)
export const SEVERITY_OPTIONS = [
  'Catastrophic',
  'Major',
  'Moderate',
  'Minor',
];

// ── Probability ───────────────────────────────────────────────────────────────
// Mirrors Probability in enums.py (GAHAR SAC 4-level scale, most → least likely)
export const PROBABILITY_OPTIONS = [
  'Frequent',
  'Occasional',
  'Uncommon',
  'Remote',
];

// ── Action Status ─────────────────────────────────────────────────────────────
// Mirrors ActionStatus in enums.py
export const ACTION_STATUSES = [
  'Pending',
  'InProgress',
  'Completed',
];

// ── Facility Types ────────────────────────────────────────────────────────────
// DB stores Arabic canonical values — mirrors FacilityType in enums.py.
// Use FACILITY_TYPE_EN (in enums.py) or the i18n key common.facility_types.*
// for display; never change these values or existing documents will break.
export const FACILITY_TYPES = [
  'مستشفى', // Hospital
  'مركز',   // Health Center
  'وحدة',   // Health Unit
];

// ── Reporter Roles ────────────────────────────────────────────────────────────
// Clinical and operational roles available to staff reporters.
export const REPORTER_ROLES = [
  'Nurse',
  'Physician',
  'Pharmacist',
  'Lab Technician',
  'Radiology Technician',
  'Administrative Staff',
  'Other',
];

// ── Event Discovery Methods ───────────────────────────────────────────────────
// Mirrors EventDiscoveryMethod in enums.py
export const EVENT_DISCOVERY_METHODS = [
  'SelfReported',
  'SupervisorReported',
  'PatientComplaint',
  'AuditFinding',
  'IncidentInvestigation',
  'Other',
];

// ── Medication Error Stages ───────────────────────────────────────────────────
// Mirrors MedicationErrorStage in enums.py
export const MEDICATION_ERROR_STAGES = [
  'Prescribing',
  'Transcribing',
  'Dispensing',
  'Preparation',
  'Administration',
  'Monitoring',
];

// ── NCC MERP Categories ───────────────────────────────────────────────────────
// Mirrors NCCMERPCategory in enums.py (A–I), enriched with display metadata.
// `severity` drives the colour bands in the reference bar and GAHAR panel.
export const NCC_MERP_CATEGORIES = [
  { value: 'A', emoji: '🟢', severity: 'safe' },      // No error occurred
  { value: 'B', emoji: '🟡', severity: 'near_miss' }, // Did not reach patient
  { value: 'C', emoji: '🟡', severity: 'near_miss' }, // Reached patient, no harm
  { value: 'D', emoji: '🟠', severity: 'near_miss' }, // Monitoring required
  { value: 'E', emoji: '🔴', severity: 'harm' },      // Temporary harm
  { value: 'F', emoji: '🔴', severity: 'harm' },      // Hospitalisation
  { value: 'G', emoji: '🟣', severity: 'severe' },    // Permanent harm
  { value: 'H', emoji: '🟣', severity: 'severe' },    // Life-sustaining intervention
  { value: 'I', emoji: '⚫', severity: 'fatal' },     // Patient death
];

// ── Occurrence Location — Facility-type-aware ─────────────────────────────────
// NewIncidentForm picks the right list based on watch('facility_type'):
//   'مستشفى' (Hospital)  → HOSPITAL_LOCATION_OPTIONS
//   'مركز' / 'وحدة'      → OTHER_FACILITY_LOCATION_OPTIONS
//
// Slugs are stored in the DB as-is and translated in the UI via:
//   t(`incidents.occurrence_location.${slug}`, { defaultValue: formatEnumLabel(slug) })
//
// Both lists share telehealth_remote and administrative_area so they always appear
// regardless of facility type.

export const HOSPITAL_LOCATION_OPTIONS = [
  'emergency_room',
  'icu',
  'operating_room',
  'recovery_room',
  'inpatient_ward',
  'outpatient_clinic',
  'radiology',
  'laboratory',
  'pharmacy',
  'maternity_ward',
  'pediatric_ward',
  'cardiac_care',
  'corridor',
  'administrative_area',
  'telehealth_remote',
];

export const OTHER_FACILITY_LOCATION_OPTIONS = [
  'examination_room',
  'treatment_room',
  'waiting_area',
  'reception',
  'nursing_station',
  'patient_room',
  'procedure_room',
  'administrative_area',
  'telehealth_remote',
];